python/lsst/daf/butler/datastores/s3Datastore.py

```diff
@@ -0,0 +1,719 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""S3 datastore."""
+
+__all__ = ("S3Datastore", )
+
+import os
+import logging
+from collections import namedtuple
+from urllib.parse import urlparse
+import tempfile
+
+import boto3
+
+from lsst.daf.butler import (Config, Datastore, DatastoreConfig, S3LocationFactory,
+                             Location, FileDescriptor, FormatterFactory, FileTemplates,
+                             StoredFileInfo, StorageClassFactory, DatasetTypeNotSupportedError,
+                             DatabaseDict, DatastoreValidationError, FileTemplateValidationError,
+                             Constraints)
+from lsst.daf.butler.core.utils import transactional, getInstanceOf
+from lsst.daf.butler.core.s3utils import (s3CheckFileExists,
+                                          parsePathToUriElements)
+from lsst.daf.butler.core.repoRelocation import replaceRoot
+
+log = logging.getLogger(__name__)
+
+
+class S3Datastore(Datastore):
+    """Basic S3 Object Storage backed Datastore.
+
+    Attributes
+    ----------
+    config : `DatastoreConfig`
+        Configuration used to create Datastore.
+    registry : `Registry`
+        `Registry` to use when recording the writing of Datasets.
+    root : `str`
+        Root directory of this `Datastore`.
+    s3locationFactory : `LocationFactory`
+        Factory for creating locations relative to S3 bucket.
+    locationFactory : `LocationFactory`
+        Factory for creating locations relative to datastore root
+        as if it was file:// .
+    formatterFactory : `FormatterFactory`
+        Factory for creating instances of formatters.
+    storageClassFactory : `StorageClassFactory`
+        Factory for creating storage class instances from name.
+    templates : `FileTemplates`
+        File templates that can be used by this `Datastore`.
+    name : `str`
+        Label associated with this Datastore.
+
+    Parameters
+    ----------
+    config : `DatastoreConfig` or `str`
+        Configuration.
+
+    Raises
+    ------
+    ValueError
+        If root location does not exist and ``create`` is `False` in the
+        configuration.
+    """
+
+    defaultConfigFile = "datastores/s3Datastore.yaml"
+    """Path to configuration defaults. Relative to $DAF_BUTLER_DIR/config or
+    absolute path. Can be None if no defaults specified.
+    """
+
+    RecordTuple = namedtuple("s3datastorerecord", ["formatter", "path",
+                                                   "storage_class", "checksum",
+                                                   "file_size"])
+
+    @classmethod
+    def setConfigRoot(cls, root, config, full, overwrite=True):
+        """Set any filesystem-dependent config options for this Datastore to
+        be appropriate for a new empty repository with the given root.
+        Parameters
+        ----------
+        root : `str`
+            Filesystem path to the root of the data repository.
```

Comment:
We probably want to change this to something like "URI to the root of the data repository"